packages/kbn-observable/src/observable.test.ts

```diff
@@ -878,6 +863,67 @@ describe('observable#pipe', () => {
     expect(actual).toEqual(['2', '3', '4']);
   });
 
+  test('should be able to unsubscribe within a subscribe', () => {
```

Comment:
This test is not consistent with RxJS, and it shows a problem with our subscription handling. Instead of the current assertions below:

```js
expect(factoryState).toEqual([
  { x: 1, isClosed: false },
  { x: 2, isClosed: false },
  { x: 3, isClosed: false },
  { x: 4, isClosed: false },
]);
expect(operator1State).toEqual([
  { x: 1, isClosed: false },
  { x: 2, isClosed: false },
  { x: 3, isClosed: true },
  { x: 4, isClosed: true },
]);
expect(operator2State).toEqual([
  { x: 1, isClosed: false },
  { x: 2, isClosed: false },
]);
expect(output).toEqual([1, 2]);
```

The results should have been:

```js
expect(factoryState).toEqual([
  { x: 1, isClosed: false },
  { x: 2, isClosed: false },
  { x: 3, isClosed: true },
]);
expect(operator1State).toEqual([
  { x: 1, isClosed: false },
  { x: 2, isClosed: false },
]);
expect(operator2State).toEqual([
  { x: 1, isClosed: false },
  { x: 2, isClosed: false },
]);
expect(output).toEqual([1, 2]);
```

i.e. when you `unsubscribe` it should close it all the way up the chain, so you can decide at the top/source how you want to handle it. When it doesn't do this it causes problems when source observable is infinite (it will likely oom, depending on what you do and the operators).

We're following the design from the TC39 proposal so I'm fairly sure that'll have the same problem too, but as no operators have been proposed there yet I expect working out those issues are part of the later work. There's a bit of complexity in RxJS around handling these subscriptions, but it could be worth looking into if you plan on pushing this forward.

We could go that route and add more complexity around this, but in the end I think it's time to consider going with RxJS 6 or some other observable library. The work they've been doing over the last 3-4 months has been great imo, and they now have Ben working full-time on it. If you want to keep exploring this, changing the assertions to what's above should ensure this case is covered. I explored passing the subscriptions around in the `subscribe`s, but it made it easy to forget doing it in operators and thereby cause subtle bugs.

Sorry, I didn't have time to set up a meeting to discuss the final parts here before I got too busy with packing up the apartment, but hopefully it gives enough context to make a decision.

/cc @elastic/kibana-platform